src/compiler/transformer.ts

```diff
@@ -316,7 +320,10 @@ namespace ts {
                 if (lexicalEnvironmentVariableDeclarations) {
                     const statement = createVariableStatement(
                         /*modifiers*/ undefined,
-                        createVariableDeclarationList(lexicalEnvironmentVariableDeclarations)
+                        createVariableDeclarationList(
+                            lexicalEnvironmentVariableDeclarations,
+                            lexicalEnvironmentScoping === LexicalEnvironmentScoping.Block ? NodeFlags.Let : undefined

```

Comment:
Previously this just a variable declaration environment, and not a block scope. I'm concerned that introducing `let` declarations here will be a problem for down-level transformations to ES5 as they won't have been visited by the type checker and won't have the necessary `NodeCheckFlags` to check for shadowing and per-iteration environments block-scoped variables down-level. This becomes even more complicated if this is part of the public API for TypeScript, as those things won't "just work" if someone builds a custom transformer.

Reply:
Thanks for pointing this out - latest changes should rely on the checker flag being set to determine if it should be a block-hoisted variable - so there's no API to say "hoist this variable into the block scope."